Subgraph updates

rithvikvibhu opened the pull request:
This PR is related to subgraphs and does a bunch of things listed below. Looks like a big PR (sorry), but some files are auto-generated.

### Use

For anyone who wants to run and query:

```sh
# start hardhat chain
yarn chain --network hardhat --hostname 0.0.0.0

# start graph node (make sure docker is running)
yarn graph-node

# deploy local subgraph (prepare creates yaml file from template, ship will compile and push to node)
# -l is version-label. Anything is fine for local, but label cannot be reused when using hosted service
yarn graph-prepare && yarn graph-ship-local -l v0.0.1
# ^ run after making changes to subgraph code

# query it!
# open http://localhost:8000/subgraphs/name/kibagateaux/xnhns-local/graphql in a browser
```

### Examples

```gql
# Get all domains
{
  domains(subgraphError: allow) {
    id
    name labelName labelhash
    parent{id} subdomains{id name labelName labelhash}
    registeredAt
    owner{id}
  }
}

# Get oracle response - like waiting for verification
{
  oracleUpdateReceiveds(subgraphError: allow, where: {domain: "0xb249cbb177b987f4423be5b672a5726d3ea6dee281efe7c4840fbabc6d38c1e7", owner: "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"}) {
    id domain{name} owner{id}
  }
}
```

I did plan to add a table of all event handlers and when they are called here, but reading **subgraph.template.yaml** and **xnhns.ts** will probably be enough.

### Changes in the PR

**packages/hardhat/**
- Reverts https://github.com/hnsfund/xnhns/commit/e3735d35196ab85779f7e7d263467b5c36356d01 because it's not possible to retrieve indexed strings [as only hashes are stored](https://ethereum.stackexchange.com/questions/71716/why-do-my-events-not-emit-the-correct-value). Instead, the input of the method call is decoded to get the tld as string.

**packages/subgraph/abis/** - these are generated ABIs for the revert above

**packages/subgraph/src/**
- **types/** - generated types based on GraphQL schema
- **subgraph.template.yaml**
  - updated apiVersion from 0.0.4 to 0.0.6 (and followed [migration guide](https://thegraph.com/docs/en/developer/assemblyscript-migration-guide/))
  - updated all event handlers (commented ones are usable, but not needed events)
  - updated description and repo url
- **ensRegistry.ts**  - migrated and fixed breaking changes
- **resolver.ts** - added DNSResolver events, migrated and fixed breaking changes
- **xnhns.ts** - working event handlers, cleaned up code, added comments/function descs
- **package.json** - updated graphprotocol dependencies to latest versions (needed for `ethereum.decode`)
- **schema.graphql** - added PublicResolver (including DNS) types

### Todo

I haven't tried snitching (`handleSnitchedOn` and `handleSnitchesGotStiches`) yet - will be easier to test once there are... tests for snitching.

rithvikvibhu committed:
Subgraph updates

rithvikvibhu commented:
Oh and since this updates deps, will need to run `yarn` to install newer version of the libs.

rithvikvibhu committed:
subgraph: remove logs

kibagateaux commented on packages/subgraph/schema.graphql:
```diff
@@ -129,3 +129,118 @@ interface ResolverEvent {
+  zonehash: Bytes!
+}
+
+type InterfaceChanged implements ResolverEvent @entity {
```
What does this do on resolvers?

rithvikvibhu replied:
Oh the resolver types are stolen from ENS: https://github.com/ensdomains/ens-subgraph/blob/master/schema.graphql and are used by `resolver.ts` which already existed before.

The 3 DNS event types (DNS*) are new and not from `ens-subgraph`.

rithvikvibhu replied:
Or was your Q more general about ResolverEvents? These are similar to DomainEvents and store changes made on different kinds of records (addr, contenthash, dns, etc.)

kibagateaux replied:
My question was about `InterfaceChanged` specifically and what that means for resolvers. I see it already existed in codebase before your PR so just curious

rithvikvibhu replied:
I think that's related to `InterfaceResolver` that `PublicResolver` inherits from: https://github.com/ensdomains/resolvers/blob/master/contracts/profiles/InterfaceResolver.sol

It's from EIP-1844 but I don't think it's used? https://eips.ethereum.org/EIPS/eip-1844

kibagateaux commented on packages/subgraph/src/ensRegistry.ts:
```diff
@@ -21,33 +19,40 @@ import {
   account.save()
 
-  let subnode = crypto.keccak256(concat(event.params.node, event.params.label)).toHexString()
+  let subnode =  makeSubnode(event)
```
Maybe rename to `namehash` instead of `subnode`? Subnode isn't a term in ENS as far as im aware but namehash represents the hash of the domain you are working with

rithvikvibhu replied:
I actually got that from `ens-subgraph` but yeah can change it to namehash.
https://github.com/ensdomains/ens-subgraph/blob/71ea385087821947f5f087fbc92bf104e43db55a/src/ensRegistry.ts#L45-L47

kibagateaux commented on packages/subgraph/src/ensRegistry.ts:
```diff
@@ -68,12 +73,11 @@ function _handleNewOwner(event: NewOwnerEvent): void {
 
   // Update the domain owner
-  let domain = getDomain(node)
+  let domain = getDomain(node)!
```
is `!` syntax a new JS thing or subgraph thing? I assume it means the return value can't be null?

rithvikvibhu replied:
Yup exactly, that's one of the breaking changes in assemblyscript: https://thegraph.com/docs/en/developer/assemblyscript-migration-guide/#nullability

kibagateaux commented on packages/subgraph/src/resolver.ts:
```diff
@@ -62,28 +68,30 @@ export function handleMulticoinAddrChanged(event: AddressChangedEvent): void {
   resolverEvent.blockNumber = event.block.number.toI32()
   resolverEvent.transactionID = event.transaction.hash
-  resolverEvent.coinType = coinType
+  resolverEvent.coinType = BigInt.fromI32(coinType)
```
Is there a reason the `coinType` is BigInt on MulticoinAddrChanged event but `coinTypes` on Resolver is just Int? Shouldn't they be the same so we don't have to convert between them at some later point?

rithvikvibhu replied:
Oh good catch. It was changed from Int to BigInt in https://github.com/ensdomains/ens-subgraph/pull/33, looks like I missed that. Will change here too.

kibagateaux commented on packages/subgraph/src/resolver.ts:
```diff
@@ -145,6 +155,47 @@ export function handleContentHashChanged(event: ContenthashChangedEvent): void {
   resolverEvent.save()
 }
 
+export function handleDNSRecordChanged(event: DNSRecordChangedEvent): void {
```
Are all these DNS functions new code not forked from ENS repo? If so i will take a deeper look at them

rithvikvibhu replied:
Yup, ens-sugraph doesn't have DNS events. If this works, then can maybe PR to their repo :p

kibagateaux replied:
Have you tried creating an API request to get DNS records for a domain? I'm playing around with the subgraph on mumbai and it seems there is no way to filter Resolver events by type and records aren't stored at top level Domain either. So we'd have to pull in all resolver events for a domain and then filter on front-end for DNS records? 

Might be worth considering creating a `Record` entity in schema and creating field `Domain.records: [Record!]`. Based on ENS setup I think Record would have to be an RRSet not an individual DNS record. Can still keep all the ResolverEvents just have them update the Record entity and save to Domain for easier access.

I could be totally wrong though, let me know what you think

rithvikvibhu replied:
DNS records are submitted to the resolver contract as multiple DNS wire format strings. The contract decodes each name and stores it by hash so there's no way to get all records (for different names) unless you specify exactly what "name" and "type" to get records for ("sh8.", "A").

So yeah, subgraphs aren't useful to extract dns records so these events are nice to have to keep track of what was added (since the full list cannot be retrieved from the resolver at all).

Same goes for Text records. Subgraphs only store keys, not values. ENS app gets these keys from the subgraph and queries the resolver to get values for each key.

rithvikvibhu replied:
About storing parsed DNS records on subgraph, would be great if possible but it's assemblyscript and not sure if bns (required to parse dns wire format) can be used in event handlers.

rithvikvibhu replied:
Thinking now, it may be possible to modify DNSResolver to emit keys (dns record names as strings and use it in the same way as how TXT records are used (prev comment)

kibagateaux replied:
I think it should be possible since everything gets built and bundled before being sent to the graph indexers to run. Like we have `@ensdomains/content-hash` package included in our subgraph code which isn't from official Graph libraries and all packages in subgraph are listed as dev dependencies. 

If not possible its not a problem, huge improvement already.

kibagateaux commented on packages/subgraph/src/resolver.ts:
```diff
@@ -62,28 +68,30 @@ export function handleMulticoinAddrChanged(event: AddressChangedEvent): void {
   resolverEvent.blockNumber = event.block.number.toI32()
   resolverEvent.transactionID = event.transaction.hash
-  resolverEvent.coinType = coinType
+  resolverEvent.coinType = BigInt.fromI32(coinType)
```
Is there a reason the `coinType` is BigInt on MulticoinAddrChanged event but coinTypes on Resolver is just Int? Shouldn't they be the same so we don't have to convert between them at some later point?

rithvikvibhu replied:
answered here: https://github.com/hnsfund/xnhns/pull/2#discussion_r780736218

rithvikvibhu committed:
coinType type and subnode -> namehash

rithvikvibhu commented:
Made changes to coinType type and renamed to namehash based on review. No errors, updated on https://thegraph.com/hosted-service/subgraph/rithvikvibhu/xnhns-mumbai also.
